core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -453,7 +462,7 @@ private[log] class Cleaner(val id: Int,
           if (writeOriginalMessageSet)
             ByteBufferMessageSet.writeMessage(writeBuffer, entry.message, entry.offset)
           else if (retainedMessages.nonEmpty)
-            compressMessages(writeBuffer, entry.message.compressionCodec, messageFormatVersion, retainedMessages)
+            compressMessages(writeBuffer, targetCodec, messageFormatVersion, retainedMessages)
```

Comment:
We probably also want to change the logic that sets `writeOriginalMessageSet` to check if source codec is the same as target codec.